Source/WebCore/dom/ContainerNodeAlgorithms.cpp

```diff
@@ -101,38 +101,58 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
         currentObservability = newStatus;
 }
 
-static RemovedSubtreeObservability notifyNodeRemovedFromDocument(ContainerNode& oldParentOfRemovedTree, TreeScopeChange treeScopeChange, Node& node)
+static RemovedSubtreeResult notifyNodeRemovedFromDocument(ContainerNode& oldParentOfRemovedTree, TreeScopeChange treeScopeChange, Node& node)
 {
     ASSERT(!node.parentNode());
     ASSERT(oldParentOfRemovedTree.isConnected());
     ASSERT(node.isConnected());
 
     RemovedSubtreeObservability observability = RemovedSubtreeObservability::NotObservable;
+    auto childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::Yes;
+    unsigned subTreeSize = 0;
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
+        ++subTreeSize;
         currentNode->removedFromAncestor(Node::RemovalType { /* disconnectedFromDocument */ true, treeScopeChange == TreeScopeChange::Changed }, oldParentOfRemovedTree);
+        if (!AsyncNodeDeletionQueue::canNodeBeDeletedAsync(node))
+            childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::No;
         updateObservability(observability, observabilityOfRemovedNode(*currentNode));
-        if (RefPtr root = currentNode->shadowRoot())
-            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+        if (RefPtr root = currentNode->shadowRoot()) {
+            auto [shadowSubTreeSize, newObservability, canBeDelayed] = notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root);
+            if (canBeDelayed == ContainerNode::ChildrenDeletionCanBeDelayed::No)
```

Comment:
We should probably add a helper function like updateChildrenDeletionCanBeDelayed to share code.

Reply:
By the way, ChildrenDeletionCanBeDelayed is quite wordy. How about CanDelayNodeDeletion or CanDeleteNodeAsync?